tests/unit/test_charm.py

```diff
@@ -45,17 +48,69 @@
 
 
-class TestCharm(unittest.TestCase):
-    def setUp(self):
-        self.maxDiff = None
-        self.namespace = "whatever"
-        self.metadata = self._get_metadata()
-        self.container_name = list(self.metadata["containers"].keys())[0]
+class TestCharm:
+    patcher_check_output = patch("charm.check_output")
+    patcher_nrf_url = patch(
+        "charms.sdcore_nrf_k8s.v0.fiveg_nrf.NRFRequires.nrf_url",
+        new_callable=PropertyMock
+    )
+    patcher_generate_csr = patch("charm.generate_csr")
+    patcher_generate_private_key = patch("charm.generate_private_key")
+    patcher_get_assigned_certificates = patch("charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.get_assigned_certificates")  # noqa: E501
+    patcher_request_certificate_creation = patch("charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.request_certificate_creation")  # noqa: E501
+    patcher_restart = patch("ops.model.Container.restart")
+
+    @pytest.fixture()
+    def setup(self):
+        self.mock_check_output = TestCharm.patcher_check_output.start()
+        self.mock_check_output.return_value = POD_IP
+        self.mock_nrf_url = TestCharm.patcher_nrf_url.start()
+        self.mock_nrf_url.return_value = VALID_NRF_URL
+        self.mock_generate_csr = TestCharm.patcher_generate_csr.start()
+        self.mock_generate_csr.return_value = STORED_CSR
+        self.mock_generate_private_key = TestCharm.patcher_generate_private_key.start()
+        self.mock_generate_private_key.return_value = PRIVATE_KEY
+        self.mock_get_assigned_certificates = TestCharm.patcher_get_assigned_certificates.start()
+        self.mock_request_certificate_creation = TestCharm.patcher_request_certificate_creation.start()  # noqa: E501
+        self.mock_restart = TestCharm.patcher_restart.start()
+
+    @staticmethod
+    def teardown() -> None:
+        patch.stopall()
+
+    @pytest.fixture(autouse=True)
+    def create_harness(self, setup, request):
         self.harness = testing.Harness(NSSFOperatorCharm)
-        self.harness.set_model_name(name=self.namespace)
-        self.addCleanup(self.harness.cleanup)
+        self.harness.set_model_name(name=NAMESPACE)
         self.harness.set_leader(is_leader=True)
         self.harness.begin()
+        yield self.harness
+        self.harness.cleanup()
+        request.addfinalizer(self.teardown)
+
+    @pytest.fixture()
+    def add_storage(self):
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.add_storage(storage_name="config", attach=True)
+
+    @pytest.fixture()
+    def fiveg_nrf_relation_id(self) -> Generator[int, None, None]:
+        relation_id = self.harness.add_relation(
+            relation_name=NRF_RELATION_NAME,
+            remote_app="whatever-nrf",
+        )
+        self.harness.add_relation_unit(
+            relation_id=relation_id, remote_unit_name="whatever-nrf/0"
+        )
+        yield relation_id
```

Comment:
does it need to yield the `relation_id` even if it is never explicitly used?
Same for `certificates_relation_id`

Reply:
The `relation_id` is used in 3 tests: `test_given_nssf_charm_in_active_status_when_nrf_relation_breaks_then_status_is_blocked`, test_given_nssf_charm_in_active_status_when_certificates_relation_breaks_then_status_is_blocked`, `test_given_container_cannot_connect_when_certificates_relation_breaks_then_waiting_for_container_to_start`.